From ab77dfaaabea44c831820cda38b386510d7e9aa9 Mon Sep 17 00:00:00 2001
From: Rabeeh Khoury <rabeeh@solid-run.com>
Date: Tue, 18 Feb 2014 12:46:07 +0200
Subject: [PATCH] Proper wifi/bt pad and pinout handling

1. bcm4330 has extra shutdown pin more than bcm4329; so handle this pin carefully.
2. bcm4330 and bcm4329 devices are connected to aud3 and not aud5 (production version microsom).

Signed-off-by: Rabeeh Khoury <rabeeh@solid-run.com>
---
 arch/arm/mach-mx6/board-mx6dl_microsom.h | 18 ++++++++----------
 arch/arm/mach-mx6/board-mx6q_microsom.c  |  4 ++++
 arch/arm/mach-mx6/board-mx6q_microsom.h  | 13 +++++++------
 3 files changed, 19 insertions(+), 16 deletions(-)

diff --git a/arch/arm/mach-mx6/board-mx6dl_microsom.h b/arch/arm/mach-mx6/board-mx6dl_microsom.h
index 2eb9eb5..206eae2 100644
--- a/arch/arm/mach-mx6/board-mx6dl_microsom.h
+++ b/arch/arm/mach-mx6/board-mx6dl_microsom.h
@@ -29,18 +29,19 @@
 	MX6DL_PAD_GPIO_8__ANATOP_ANATOP_32K_OUT,
 	MX6DL_PAD_CSI0_DAT8__GPIO_5_26,		/* WL_RST_IN */
 	MX6DL_PAD_CSI0_DAT9__GPIO_5_27,		/* WL_GPIO_0 */
-	MX6DL_PAD_CSI0_DATA_EN__GPIO_5_20,		/* WL_GPIO_1 */
+	MX6DL_PAD_CSI0_DATA_EN__GPIO_5_20,	/* WL_GPIO_1 */
 	MX6DL_PAD_EIM_D19__GPIO_3_19,		/* WL_BT_REG_ON */
-	MX6DL_PAD_DISP0_DAT21__GPIO_5_15,	/* XTAL power up */
+	MX6DL_PAD_DISP0_DAT11__GPIO_5_5,        /* XTAL power up */
 	MX6DL_PAD_CSI0_DAT14__GPIO_6_0,		/* BT_RST_N */
 	MX6DL_PAD_CSI0_DAT15__GPIO_6_1,		/* BT_GPIO_0 */
 	MX6DL_PAD_CSI0_DAT18__GPIO_6_4,		/* BT_GPIO_1 */
+        MX6DL_PAD_CSI0_DAT19__GPIO_6_5,		/* WL_SHUTDOWN_N - BRCM 4330 only */
 
 	/* AUD5 for BRCM WiFi/BT/FM */
-	MX6DL_PAD_KEY_COL0__AUDMUX_AUD5_TXC,
-	MX6DL_PAD_KEY_ROW0__AUDMUX_AUD5_TXD,
-	MX6DL_PAD_KEY_COL1__AUDMUX_AUD5_TXFS,
-	MX6DL_PAD_KEY_ROW1__AUDMUX_AUD5_RXD,
+	MX6DL_PAD_CSI0_DAT4__AUDMUX_AUD3_TXC,
+	MX6DL_PAD_CSI0_DAT5__AUDMUX_AUD3_TXD,
+	MX6DL_PAD_CSI0_DAT6__AUDMUX_AUD3_TXFS,
+	MX6DL_PAD_CSI0_DAT7__AUDMUX_AUD3_RXD,
 
 	/* USDHC1 - Connected to BRCM Wifi/BT/FM */
 	MX6DL_PAD_SD1_CLK__USDHC1_CLK_50MHZ_40OHM,
@@ -106,9 +107,9 @@
 	MX6DL_PAD_ENET_MDIO__ENET_MDIO,
 	MX6DL_PAD_ENET_MDC__ENET_MDC,
 	IOMUX_PAD(0x0650, 0x0268, 5, 0x0000, 0, MX6DL_ENET_PAD_CTRL_PD), /* KEY_ROW4 reset signal */
-	
 	MX6DL_PAD_DI0_PIN2__GPIO_4_18, /* Interrupt */
 	/* RMII */
+
 	IOMUX_PAD(0x05B4, 0x01E4, 1, 0x0828, 0, PAD_CTL_PKE | PAD_CTL_PUE  |
 		  PAD_CTL_PUS_100K_DOWN), /* MX6DL_PAD_ENET_CRS_DV__ENET_RX_EN */
 	IOMUX_PAD(0x05C8, 0x01F8, 1, 0x0818, 0, PAD_CTL_PKE | PAD_CTL_PUE  |
@@ -136,8 +137,6 @@
 	IOMUX_PAD(0x06A4, 0x02BC, 1, 0x0828, 1, MX6DL_ENET_PAD_CTRL_PD),/*RGMII RXCTL*/
 };
 
-
-
 static iomux_v3_cfg_t mx6dl_usom_hdmi_ddc_pads[] = {
 	MX6DL_PAD_KEY_COL3__HDMI_TX_DDC_SCL, /* HDMI DDC SCL */
 	MX6DL_PAD_KEY_ROW3__HDMI_TX_DDC_SDA, /* HDMI DDC SDA */
@@ -147,5 +146,4 @@
 	MX6DL_PAD_KEY_COL3__I2C2_SCL,	/* I2C2 SCL */
 	MX6DL_PAD_KEY_ROW3__I2C2_SDA,	/* I2C2 SDA */
 };
-
 #endif
diff --git a/arch/arm/mach-mx6/board-mx6q_microsom.c b/arch/arm/mach-mx6/board-mx6q_microsom.c
index e631293..a6d4936 100644
--- a/arch/arm/mach-mx6/board-mx6q_microsom.c
+++ b/arch/arm/mach-mx6/board-mx6q_microsom.c
@@ -87,6 +87,7 @@
 
 #define USOM_WL_RST		IMX_GPIO_NR(5, 26)
 #define USOM_BT_RST		IMX_GPIO_NR(6, 0)
+#define USOM_BT_SHUTDOWN	IMX_GPIO_NR(6,15)
 #define USOM_REG_ON		IMX_GPIO_NR(3, 19)
 #define USOM_XTAL_ON		IMX_GPIO_NR(5, 5)
 #define USOM_ENET_RST		IMX_GPIO_NR(4, 15)
@@ -580,6 +581,8 @@ static void __init imx6q_usom_init_wifi_bt(void) {
 	gpio_direction_output(USOM_WL_RST, 0);
 	gpio_request(USOM_BT_RST, "bt-rst");
 	gpio_direction_output(USOM_BT_RST, 0);
+	gpio_request(USOM_BT_SHUTDOWN, "bt-shutdown");
+	gpio_direction_output(USOM_BT_SHUTDOWN, 0);
 	gpio_request(USOM_REG_ON, "wl-bt-reg-on");
 	gpio_direction_output(USOM_REG_ON, 0);
 	gpio_request(USOM_XTAL_ON, "wl-bt-xtal-on");
@@ -593,6 +596,7 @@ static void __init imx6q_usom_init_wifi_bt(void) {
 	msleep (2); /* 20 mSec sounds too big */
 	gpio_set_value(USOM_WL_RST, 1);
 	gpio_set_value(USOM_BT_RST, 1);
+	gpio_set_value(USOM_BT_SHUTDOWN, 1);
 	msleep (2); /* 20 mSec sounds too big */
 
 	/* Register SDIO as brfmac */
diff --git a/arch/arm/mach-mx6/board-mx6q_microsom.h b/arch/arm/mach-mx6/board-mx6q_microsom.h
index 482f2c1..a348e5a 100644
--- a/arch/arm/mach-mx6/board-mx6q_microsom.h
+++ b/arch/arm/mach-mx6/board-mx6q_microsom.h
@@ -29,18 +29,19 @@
 	MX6Q_PAD_GPIO_8__ANATOP_ANATOP_32K_OUT,
 	MX6Q_PAD_CSI0_DAT8__GPIO_5_26,		/* WL_RST_IN */
 	MX6Q_PAD_CSI0_DAT9__GPIO_5_27,		/* WL_GPIO_0 */
-	MX6Q_PAD_CSI0_DATA_EN__GPIO_5_20,		/* WL_GPIO_1 */
+	MX6Q_PAD_CSI0_DATA_EN__GPIO_5_20,	/* WL_GPIO_1 */
 	MX6Q_PAD_EIM_D19__GPIO_3_19,		/* WL_BT_REG_ON */
-	MX6Q_PAD_DISP0_DAT21__GPIO_5_15,	/* XTAL power up */
+	MX6Q_PAD_DISP0_DAT11__GPIO_5_5,         /* XTAL power up */
 	MX6Q_PAD_CSI0_DAT14__GPIO_6_0,		/* BT_RST_N */
 	MX6Q_PAD_CSI0_DAT15__GPIO_6_1,		/* BT_GPIO_0 */
 	MX6Q_PAD_CSI0_DAT18__GPIO_6_4,		/* BT_GPIO_1 */
+        MX6Q_PAD_CSI0_DAT19__GPIO_6_5,		/* WL_SHUTDOWN_N - BRCM 4330 only */
 
 	/* AUD5 for BRCM WiFi/BT/FM */
-	MX6Q_PAD_KEY_COL0__AUDMUX_AUD5_TXC,
-	MX6Q_PAD_KEY_ROW0__AUDMUX_AUD5_TXD,
-	MX6Q_PAD_KEY_COL1__AUDMUX_AUD5_TXFS,
-	MX6Q_PAD_KEY_ROW1__AUDMUX_AUD5_RXD,
+	MX6Q_PAD_CSI0_DAT4__AUDMUX_AUD3_TXC,
+	MX6Q_PAD_CSI0_DAT5__AUDMUX_AUD3_TXD,
+	MX6Q_PAD_CSI0_DAT6__AUDMUX_AUD3_TXFS,
+	MX6Q_PAD_CSI0_DAT7__AUDMUX_AUD3_RXD,
 
 	/* USDHC1 - Connected to BRCM Wifi/BT/FM */
 	MX6Q_PAD_SD1_CLK__USDHC1_CLK_50MHZ_40OHM,
-- 
1.8.5.5

